experimental/css-has-pseudo/src/browser.js

```diff
@@ -0,0 +1,178 @@
+/* global MutationObserver,requestAnimationFrame,cancelAnimationFrame,self */
+
+import '@mrhenry/core-web/modules/~element-qsa-has.js';
+import extractEncodedSelectors from './encode/extract.mjs';
+
+export default function cssHasPseudo(document, options) {
+	// OPTIONS
+	{
+		if (!options) {
+			options = {};
+		}
+
+		options = {
+			hover: (!!options.hover) || false,
+			debug: (!!options.debug) || false,
+			observedAttributes: options.observedAttributes || [],
+		};
+
+		if (!Array.isArray(options.observedAttributes)) {
+			options.observedAttributes = [];
+		}
+
+		options.observedAttributes = options.observedAttributes.filter((x) => {
+			return (typeof x === 'string');
+		});
+
+		// https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes
+		// `data-*` and `style` were omitted
+		options.observedAttributes = options.observedAttributes.concat(['accept', 'accept-charset', 'accesskey', 'action', 'align', 'allow', 'alt', 'async', 'autocapitalize', 'autocomplete', 'autofocus', 'autoplay', 'buffered', 'capture', 'challenge', 'charset', 'checked', 'cite', 'class', 'code', 'codebase', 'cols', 'colspan', 'content', 'contenteditable', 'contextmenu', 'controls', 'coords', 'crossorigin', 'csp', 'data', 'datetime', 'decoding', 'default', 'defer', 'dir', 'dirname', 'disabled', 'download', 'draggable', 'enctype', 'enterkeyhint', 'for', 'form', 'formaction', 'formenctype', 'formmethod', 'formnovalidate', 'formtarget', 'headers', 'hidden', 'high', 'href', 'hreflang', 'http-equiv', 'icon', 'id', 'importance', 'integrity', 'intrinsicsize', 'inputmode', 'ismap', 'itemprop', 'keytype', 'kind', 'label', 'lang', 'language', 'list', 'loop', 'low', 'manifest', 'max', 'maxlength', 'minlength', 'media', 'method', 'min', 'multiple', 'muted', 'name', 'novalidate', 'open', 'optimum', 'pattern', 'ping', 'placeholder', 'poster', 'preload', 'radiogroup', 'readonly', 'referrerpolicy', 'rel', 'required', 'reversed', 'rows', 'rowspan', 'sandbox', 'scope', 'scoped', 'selected', 'shape', 'size', 'sizes', 'slot', 'span', 'spellcheck', 'src', 'srcdoc', 'srclang', 'srcset', 'start', 'step', 'summary', 'tabindex', 'target', 'title', 'translate', 'type', 'usemap', 'value', 'width', 'wrap']);
+	}
+
+	const observedItems = [];
+
+	// document.createAttribute() doesn't support `:` in the name. innerHTML does
+	const attributeElement = document.createElement('x');
+
+	// walk all stylesheets to collect observed css rules
+	[].forEach.call(document.styleSheets, walkStyleSheet);
+	transformObservedItemsThrottled();
+
+	// observe DOM modifications that affect selectors
+	if ('MutationObserver' in self) {
```

Comment:
Given this is required by the requisites, I think we should test this earlier on, and if not just throw and not run. 

I don't see (but could be wrong) if we're checking support for `:has` on JS before allowing any of this code to run.

Reply:
Mutation observer is actually not a requirement for this polyfill.
I was surprised by this too.

It is needed to re-calc styles if html changes through client side javascript.

It's better not to throw or do early returns as this polyfill should embrace CSS concepts -> it should degrade very gracefully.

Reply:
Good to know and surprising indeed.

Thing is, as it is right now I think (if Safari was fully compliant) it would run in Safari and it wouldn't be needed?

Reply:
The polyfill would still be needed if people did not set `preserve : true`.
Since we don't know if they set `preserve : true` we have to always run the polyfill :/

Reply:
I added a new option `forcePolyfill`.
If this is not set we do feature detection and early return.

We can slim down the amount of options and config later in the experimental track.
Having more is ok now I think :)